File: LeetCode/Python/0547_Number_of_Provinces.py
# 547. Number of Provinces
# https://leetcode.com/problems/number-of-provinces/


class Solution:
    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        n = len(isConnected)
        visited = [False] * n

        def dfs(city: int):
            for neighbor in range(n):
                if isConnected[city][neighbor] == 1 and not visited[neighbor]:
                    visited[neighbor] = True
                    dfs(neighbor)

        provinces = 0

        for i in range(n):
            if not visited[i]:
                visited[i] = True
                dfs(i)
                provinces += 1

        return provinces

File: LeetCode/JavaScript/0547-Number-of-Provinces.js
// 547. Number of Provinces
// https://leetcode.com/problems/number-of-provinces/submissions/1905393463/?envType=study-plan-v2&envId=leetcode-75

/**
 * @param {number[][]} isConnected
 * @return {number}
 */

var findCircleNum = function (isConnected) {
  const n = isConnected.length;
  const visited = new Array(n).fill(false);

  function dfs(city) {
    for (let neighbor = 0; neighbor < n; neighbor++) {
      if (isConnected[city][neighbor] === 1 && !visited[neighbor]) {
        visited[neighbor] = true;
        dfs(neighbor);
      }
    }
  }

  let provinces = 0;

  for (let i = 0; i < n; i++) {
    if (!visited[i]) {
      visited[i] = true;
      dfs(i);
      provinces++;
    }
  }

  return provinces;
};
